src/System.Net.Sockets/tests/FunctionalTests/ReceiveMessageFrom.cs

```diff
@@ -77,74 +79,116 @@ public void Success_IPv6()
-                    sender.Bind(new IPEndPoint(IPAddress.IPv6Loopback, 0));
-
-                    for (int i = 0; i < TestSettings.UDPRedundancy; i++)
-                    {
-                        sender.SendTo(new byte[1024], new IPEndPoint(IPAddress.IPv6Loopback, port));
-                    }
+                if (!Socket.OSSupportsIPv4) return;
+                family = AddressFamily.InterNetwork;
+                loopback = IPAddress.Loopback;
+                any = IPAddress.Any;
+                level = SocketOptionLevel.IP;
+            }
+            else
+            {
+                if (!Socket.OSSupportsIPv6) return;
+                family = AddressFamily.InterNetworkV6;
+                loopback = IPAddress.IPv6Loopback;
+                any = IPAddress.IPv6Any;
+                level = SocketOptionLevel.IPv6;
+            }
 
-                    IPPacketInformation packetInformation;
-                    SocketFlags flags = SocketFlags.None;
-                    EndPoint remoteEP = new IPEndPoint(IPAddress.IPv6Any, 0);
+            using (var receiver = new Socket(family, SocketType.Dgram, ProtocolType.Udp))
+            using (var sender = new Socket(family, SocketType.Dgram, ProtocolType.Udp))
+            using (var saea = new SocketAsyncEventArgs())
+            {
+                int port = receiver.BindToAnonymousPort(loopback);
+                receiver.SetSocketOption(level, SocketOptionName.PacketInformation, true);
+                sender.Bind(new IPEndPoint(loopback, 0));
 
-                    IAsyncResult ar = receiver.BeginReceiveMessageFrom(new byte[1024], 0, 1024, flags, ref remoteEP, null, null);
-                    ar.AsyncWaitHandle.WaitOne();
-                    int len = receiver.EndReceiveMessageFrom(ar, ref flags, ref remoteEP, out packetInformation);
+                saea.RemoteEndPoint = new IPEndPoint(any, 0);
+                if (multiBuffers)
+                {
+                    saea.BufferList = new List<ArraySegment<byte>> { new ArraySegment<byte>(new byte[1024]) };
+                }
+                else
+                {
+                    saea.SetBuffer(new byte[1024], 0, 1024);
+                }
 
-                    Assert.Equal(1024, len);
-                    Assert.Equal(sender.LocalEndPoint, remoteEP);
-                    Assert.Equal(((IPEndPoint)sender.LocalEndPoint).Address, packetInformation.Address);
+                var mres = new ManualResetEventSlim();
+                saea.Completed += delegate { mres.Set(); };
 
-                    sender.Dispose();
+                bool pending = receiver.ReceiveMessageFromAsync(saea);
+                for (int i = 0; i < TestSettings.UDPRedundancy; i++)
+                {
+                    sender.SendTo(new byte[1024], new IPEndPoint(loopback, port));
                 }
+                if (pending) Assert.True(mres.Wait(30000), "Expected operation to complete within timeout");
```

Comment:
For operations that we expect to complete basically instantly, we've typically used a large timeout, to account for possible variation/load in a CI environment... if it ends up taking 30s, something is very wrong, rather than it being a fluke of the environment.